src/preload.py

```diff
@@ -42,116 +45,371 @@
+                    seek=part.start_bytes,
+                    count=part.size_bytes,
+                )
+        # Copy partitions after.
+        for part in partition_table.partitions[self.number:]:
+            if not part.is_extended():
+                copy(
+                    skip=part.start_bytes,
+                    seek=part.start_bytes + additional_bytes,
+                    count=part.size_bytes,
+                )
+        # Replace the original image contents.
+        ddd(_if=tmp.name, of=image, bs=1024 ** 2)
 
+    def resize(self, additional_bytes):
+        # Is it the last partition on the disk?
+        if self.is_last():
+            self._resize_last_partition_of_disk_image(additional_bytes)
+        else:
+            self._resize_partition_on_disk_image(additional_bytes)
+        expand_filesystem(self)
+
+
+class PartitionTable(object):
+    def __init__(self, image):
+        self.image = image
+        data = json.loads(
+            sfdisk("--dump", "--json", image).stdout.decode("utf8")
+        )["partitiontable"]
+        self.label = data["label"]
+        assert self.label in ("dos", "gpt")
+        self.id = data["id"]
+        self.device = data["device"]
+        self.unit = data["unit"]
+        self.firstlba = data.get("firstlba")
+        self.lastlba = data.get("lastlba")
+        self.partitions = []
+        extended_partition = None
+        for number, partition_data in enumerate(data["partitions"], 1):
+            part = Partition(self, number, **partition_data)
+            if part.is_extended():
+                extended_partition = part
+            if extended_partition and part.is_included_in(extended_partition):
+                part.set_parent(extended_partition)
+            self.partitions.append(part)
+
+    def get_partitions_in_disk_order(self):
+        # Returns the partitions in the same order that they are on the disk
+        # This excludes extended partitions.
+        partitions = (p for p in self.partitions if not p.is_extended())
+        return sorted(partitions, key=lambda p: p.start)
+
+    def get_sfdisk_script(self):
+        result = (
+            "label: {}\n"
+            "label-id: {}\n"
+            "device: {}\n"
+            "unit: {}\n"
+        ).format(self.label, self.id, self.device, self.unit)
+        if self.firstlba is not None:
+            result += "first-lba: {}\n".format(self.firstlba)
+        if self.lastlba is not None:
+            result += "last-lba: {}\n".format(self.lastlba)
+        result += "\n"
+        result += "\n".join(p.get_sfdisk_line() for p in self.partitions)
+        return result
+
+
+def get_filesystem(device):
+    line = fsck("-N", device).stdout.decode("utf8").strip().split("\n")[1]
+    return line.rsplit(" ", 2)[-2].split(".")[1]
+
+
+def expand_filesystem(partition):
+    with partition.losetup_context_manager() as device:
+        # Detects the partition filesystem (ext{2,3,4} or btrfs) and uses the
```

Comment:
Should this raise some kind of error if the partition _isn't_ using "(ext{2,3,4} or btrfs)" filesystem?

Reply:
Good idea but this PR is already merged, I'll try to think of it for the next one.